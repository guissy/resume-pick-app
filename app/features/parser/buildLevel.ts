export default function buildLevel(
  workAge: number,
  score: number,
  text: string
) {
  let lv = '-';
  let ageLv = 0;
  let scoreLv = 0;
  if (workAge >= 0 && workAge <= 3) {
    ageLv = 1;
  } else if (workAge >= 2 && workAge <= 5) {
    ageLv = 2;
  } else if (workAge >= 4 && workAge <= 7) {
    ageLv = 3;
  } else if (workAge >= 6) {
    ageLv = 4;
  }
  if (score > 2 && score <= 8) {
    scoreLv = 1;
  } else if (score >= 6 && score <= 12) {
    scoreLv = 2;
  } else if (score >= 10 && score <= 16) {
    scoreLv = 3;
  } else if (score >= 14) {
    scoreLv = 4;
  }
  if (ageLv + scoreLv >= 8) {
    lv = 'p7+';
  } else if (ageLv + scoreLv === 7) {
    lv = 'p7';
  } else if (ageLv + scoreLv === 6) {
    lv = 'p6+';
  } else if (ageLv + scoreLv === 5) {
    lv = 'p6';
  } else if (ageLv + scoreLv === 4) {
    lv = 'p5+';
  } else if (ageLv + scoreLv === 3) {
    lv = 'p5';
  } else if (ageLv + scoreLv === 2) {
    lv = 'p4+';
  } else if (ageLv + scoreLv === 1) {
    lv = 'p4';
  } else if (text.length > 0) {
    lv = 'p3';
  }
  return lv;
}
