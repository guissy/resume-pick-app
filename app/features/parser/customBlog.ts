const UNKNOWN_VERSION = null;
const blogs = {
  Next: {
    id: 'next',
    icon: 'next',
    url: 'https://nextjs.org/',
    npm: 'next',
    test(win: {
      __NEXT_DATA__: { buildId: unknown };
      next: { version: string };
    }) {
      // eslint-disable-next-line no-underscore-dangle
      if (win.__NEXT_DATA__ && win.__NEXT_DATA__.buildId) {
        return {
          version: (win.next && win.next.version) || UNKNOWN_VERSION,
        };
      }
      return false;
    },
  },
  Nuxt: {
    id: 'nuxt',
    icon: 'nuxt',
    url: 'https://nuxtjs.org/',
    npm: 'nuxt',
    test(win: { __NUXT__: { data: unknown }; $nuxt: unknown }) {
      // eslint-disable-next-line no-underscore-dangle
      if ((win.__NUXT__ && win.__NUXT__.data != null) || win.$nuxt) {
        return { version: UNKNOWN_VERSION };
      }
      return false;
    },
  },
  Hexo: {
    id: 'Hexo',
    icon: 'Hexo',
    npm: null,
    test() {
      const hasAPILinkElem = !!document.querySelector('meta[content*="Hexo"]');
      const hasWPIncludes = !!document.querySelectorAll(
        'meta[name=generator][content^="Hexo"]'
      ).length;

      if (!hasAPILinkElem && !hasWPIncludes) return false;
      const generatorMeta = document.querySelector(
        'meta[name=generator][content^="Hexo"]'
      );
      const version = generatorMeta
        ? generatorMeta.getAttribute('content')?.replace(/^\w+\s/, '')
        : UNKNOWN_VERSION;
      return { version };
    },
  },
  WordPress: {
    id: 'wordpress',
    icon: 'wordpress',
    url: 'https://wordpress.org/',
    npm: null,
    test() {
      const hasAPILinkElem = !!document.querySelector(
        'link[rel="https://api.w.org/"]'
      );
      const hasWPIncludes = !!document.querySelectorAll(
        'link[href*="wp-includes"], script[src*="wp-includes"]'
      ).length;

      if (!hasAPILinkElem && !hasWPIncludes) return false;

      const generatorMeta = document.querySelector(
        'meta[name=generator][content^="WordPress"]'
      );
      const version = generatorMeta
        ? generatorMeta.getAttribute('content')?.replace(/^\w+\s/, '')
        : UNKNOWN_VERSION;
      return { version };
    },
  },
  jekyll: {
    id: 'jekyll',
    test() {
      const hasAPILinkElem = !!document.querySelector(
        'meta[name=generator][content^="Jekyll"]'
      );

      if (!hasAPILinkElem) return false;

      const generatorMeta = document.querySelector(
        'meta[name=generator][content^="Jekyll"]'
      );
      const version = generatorMeta
        ? generatorMeta.getAttribute('content')?.replace(/^\w+\s/, '')
        : UNKNOWN_VERSION;
      return { version };
    },
  },
  Hugo: {
    id: 'Hugo',
    test() {
      const hasAPILinkElem = !!document.querySelector(
        'meta[name=generator][content^="Hugo"]'
      );

      if (!hasAPILinkElem) return false;

      const generatorMeta = document.querySelector(
        'meta[name=generator][content^="Hugo"]'
      );
      const version = generatorMeta
        ? generatorMeta.getAttribute('content')?.replace(/^\w+\s/, '')
        : UNKNOWN_VERSION;
      return { version };
    },
  },
  Django: {
    id: 'Django',
    test() {
      const hasAPILinkElem = !!document.querySelector(
        'a[rel="nofollow"][href="https://www.djangoproject.com/"]'
      );

      const hasOg = !!document.querySelector(
        '[prefix="og: http://ogp.me/ns# fb: http://ogp.me/ns/fb# article: http://ogp.me/ns/article#"]'
      );

      if (!hasAPILinkElem || !hasOg) return false;
      return { version: UNKNOWN_VERSION };
    },
  },
  Drupal: {
    id: 'drupal',
    icon: 'drupal',
    url: 'https://www.drupal.org/',
    npm: null,
    test(win: { Drupal?: { behaviors: string } }) {
      const generatorMeta = document.querySelector(
        'meta[name="generator"][content^="Drupal"]'
      );
      const version = generatorMeta
        ? generatorMeta.getAttribute('content')?.replace(/\D+/gi, '')
        : UNKNOWN_VERSION;

      // Detect Drupal resources patterns
      const resDrupal = /\/sites\/(?:default|all)\/(?:themes|modules|files)/;
      const res = Array.from(
        document.querySelectorAll<HTMLScriptElement & StyleSheet>(
          'link,style,script'
        ) || []
      );

      if (
        res.some((s) => resDrupal.test(s.src)) ||
        res.some((s) => resDrupal.test(s.href || '')) ||
        generatorMeta ||
        (win.Drupal && win.Drupal.behaviors)
      ) {
        return { version };
      }

      return false;
    },
  },
  TYPO3: {
    id: 'typo3',
    icon: 'typo3',
    url: 'https://typo3.org/',
    npm: null,
    test() {
      const generatorMeta = document.querySelector(
        'meta[name="generator"][content^="TYPO3"]'
      );

      // TYPO3 resource patterns / paths - search in link, style or script tags
      const resourcesTYPO3 = /\/(typo3conf|typo3temp|fileadmin)/;
      const res = Array.from(
        document.querySelectorAll<HTMLScriptElement & StyleSheet>(
          'link,style,script'
        ) || []
      );

      if (
        generatorMeta ||
        res.some((s) => resourcesTYPO3.test(s.src)) ||
        res.some((s) => resourcesTYPO3.test(s.href || ''))
      ) {
        // No version exposure available in TYPO3 due to information disclosure
        return { version: UNKNOWN_VERSION };
      }

      return false;
    },
  },
};
