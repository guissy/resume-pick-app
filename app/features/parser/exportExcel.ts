import fs from 'fs';
import ExcelJS, { Column } from 'exceljs';
import { remote } from 'electron';
import dayjs from 'dayjs';
import { Config, ScoreFile } from '../type';

export default async function exportExcel(config: Config, scores: ScoreFile[]) {
  const conf = {
    cols: [
      ['文件', 42],
      ['电话', 20],
      ['毕业院校', 25],
      ['学历', 10],
      ['期望薪水', 15],
      ['经验', 10],
      ['关键字分', 10],
      ['评级分', 10],
      ['评级', 10],
      ['性价比', 10],
      ...config[0].children.map((w) => [
        w.name,
        Math.max(w.name.length * 1.6, 10),
      ]),
      ['文件路径', 62],
    ],
    rows: scores.map((s) => [
      s.name,
      s.phone,
      s.school,
      s.degree,
      s.salary,
      Math.round((s.workAge || 0) * 2) / 2,
      parseFloat(s.score.toFixed(1)),
      parseFloat(s.levelValue.toFixed(1)),
      s.level,
      parseFloat(s.levelSalary.toFixed(2)),
      ...config[0].children.map((w) => {
        const found = s.keywords[0].children.find((w2) => w.name === w2.name);
        return found ? parseFloat(found.gained.toFixed(1)) : 0;
      }),
      s.path,
    ]),
  };
  const workbook = new ExcelJS.Workbook();
  const worksheet = workbook.addWorksheet('Resume Pick');
  worksheet.columns = conf.cols.map(
    ([header, width]) => ({ header, width } as Column)
  );
  worksheet.addRows(conf.rows);
  const row = worksheet.getRow(1);
  row.eachCell((cell) => {
    cell.alignment = { vertical: 'middle', horizontal: 'center' };
    cell.font = { name: 'Arial Black', bold: true, size: 14 };
  });
  conf.cols.forEach((_r, i) => {
    const col = worksheet.getColumn(i + 1);
    col.eachCell({ includeEmpty: true }, (cell) => {
      cell.font = { name: 'Arial', size: 14 };
    });
  });
  conf.rows
    .map((_r, i) => 'ABCDEFG'.split('').map((c) => `${c}${i + 2}`))
    .flat()
    .forEach((pos) => {
      const cell = worksheet.getCell(pos);
      cell.alignment = { vertical: 'middle', horizontal: 'center' };
      cell.font = { name: 'Arial', size: 14 };
    });
  const day = dayjs().format('YYYY-MM-DD');
  const filePath = await remote.dialog
    .showSaveDialog({
      defaultPath: `简历评估 ${day}.xlsx`,
      filters: [{ name: 'xls Files', extensions: ['xlsx'] }],
      properties: [],
    })
    .then((file) => {
      if (!file.canceled && file.filePath) {
        return file.filePath.toString();
      }
      return '';
    })
    .catch(() => {
      // console.log(err);
    });
  if (filePath) {
    workbook.xlsx
      .writeBuffer()
      .then((buf) => {
        return fs.writeFileSync(filePath, buf, 'binary');
      })
      .catch(() => {
        // console.log(err);
      });
  }
}
