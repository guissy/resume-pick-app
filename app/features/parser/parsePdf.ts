import { PDFPageProxy } from 'pdfjs-dist';

const pdfjs = require('pdfjs-dist');

pdfjs.GlobalWorkerOptions.workerSrc =
  'https://cdn.jsdelivr.net/npm/pdfjs-dist@2.5.207/build/pdf.worker.min.js';

function renderPage(pageData: PDFPageProxy) {
  return pageData.getTextContent().then((textContent) => {
    let lastY = -1;
    let text = '';
    // eslint-disable-next-line no-restricted-syntax
    for (const item of textContent.items) {
      if (lastY === item.transform[5] || !lastY) {
        text += item.str;
      } else {
        text += `\n${item.str}`;
      }
      // eslint-disable-next-line prefer-destructuring
      lastY = item.transform[5];
    }
    return text;
  });
}

export default async function parsePdf(url: string) {
  const pdf = await pdfjs.getDocument(url).promise;
  const texts = await Promise.allSettled(
    Array(pdf.numPages)
      .fill(0)
      .map((_v, i) => pdf.getPage(i + 1).then(renderPage))
  );
  return texts
    .map((v) => ((v as unknown) as { value: string }).value || '')
    .join('\n\n');
}
