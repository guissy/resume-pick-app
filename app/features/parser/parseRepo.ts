import git from 'isomorphic-git';
import http from 'isomorphic-git/http/node';
import path from 'path';
import fs from 'fs';
import { promisify } from 'util';
import { exec } from 'child_process';
import { remote } from 'electron';

export const resourcesPath = remote.app.isPackaged
  ? path.dirname(remote.app.getAppPath())
  : path.resolve('resources');

async function cloc(folder: string) {
  const { stdout } = await promisify(exec)(
    [path.join(resourcesPath, '.bin/cloc'), '--json', folder].join(' ')
  );
  return JSON.parse(stdout || '');
}

export default async function parseRepo(
  repoUrl: string,
  dir: string,
  setStatus: (info: string) => void
) {
  setStatus('git clone...');
  await git.clone({
    fs,
    http,
    dir,
    url: repoUrl,
    singleBranch: false,
    depth: 99,
  });
  const commits = await git.log({ fs, dir });
  const logs = commits
    .map((v) => v.commit.message.trim() || '')
    .filter((v) => !v.startsWith('Merge '));
  setStatus(`commits ${commits.length}`);
  const folders = fs
    .readdirSync(dir, { withFileTypes: true })
    .filter((dirent) => dirent.isDirectory())
    .filter((dirent) => !dirent.name.startsWith('.'))
    .filter((dirent) =>
      ['src', 'app', 'App', 'front', 'web', 'server', 'client'].includes(
        dirent.name
      )
    )
    .map((dirent) => dirent.name);
  const langCount = await cloc(
    path.join(dir, folders[0] || dir)
  ).catch(() => ({}));
  const fileTypes = Object.keys(langCount)
    .filter(
      (k) => !['JSON', 'Markdown', 'YAML', 'XML', 'SVG', 'header'].includes(k)
    )
    .map((lang) => langCount[lang])
    .filter(Boolean)
    .filter((v) => v.code > 50);
  const nFiles = fileTypes.map((v) => v.nFiles).reduce((s, v) => s + v, 0);
  const nComment = fileTypes.map((v) => v.comment).reduce((s, v) => s + v, 0);
  const nCode = fileTypes.map((v) => v.code).reduce((s, v) => s + v, 0);
  const commentRate = nComment / nCode;
  const linesInFile = Math.round(nCode / nFiles); // 行数/文件
  const linesInCommit = Math.round(nCode / logs.length); // 行数/提交
  setStatus(`commentRate ${(commentRate * 100).toFixed(2)}`);
  setStatus(`linesInFile ${linesInFile}`);
  setStatus(`linesInCommit ${linesInCommit}`);
  return {
    commits,
    nCode,
    commentRate,
    linesInFile,
    linesInCommit,
    fileTypes,
  };
}
