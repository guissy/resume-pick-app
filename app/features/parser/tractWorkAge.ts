import Sentiment from 'sentiment';
import { Config } from '../type';

export default function trackWorkAge(text: string) {
  // 工作年限：3年
  let years = '';
  if (text) {
    const match1 = text.match(
      // 8年前端生涯, 3年react, 3年ts, 1年RN移动端、小程序、Node服务端丰富项目经验掌握 nodejs，gql，
      /(工作)?(年限|经验|开发)[\s\t\-：:]*(([0-9一二三四五六七八九十]){1,2})\s*年/
    );
    years = match1 ? String(match1.pop()) : '';
  }
  if (!years && text) {
    // 12年工作经验 | 其他 | 38岁 | 男
    const match2 = text.match(
      /(([0-9一二三四五六七八九十]){1,2})\s*年(工作)?(年限|经验|开发)/
    );
    years = match2 ? String(match2[1]) : '';
  }
  if (!years && text) {
    const match2 = text.match(/\\b(([0-9一二三四五六七八九十]){1,2})\s*年/);
    years = match2 ? String(match2[1]) : '';
  }
  return parseInt(years, 10);
}

export function trackPhone(text: string) {
  return (text || '').match(/1\d{10}/g)?.[0] || '';
}

export function trackSchool(text: string) {
  // eslint-disable-next-line no-control-regex
  const regExp = /((?<![是的不在有这个来到时着和年就那要出也得里后自以会可下而过去能对由负责])[^\x00-\xff：，。经历]){2,8}(学院|大学|职院|师范|职中|高中|中学|一中|职业技术学校)/;
  return (text || '').match(regExp)?.[0] || '';
}

export function trackDegree(text: string) {
  const regExp = /(博士|硕士|本科|一本|二本|大专|高中)/;
  return (text || '').match(regExp)?.[0] || '';
}

export function trackSalary(text: string) {
  let regExp = /(?<=期望(薪水|薪资|工资|月薪)[:：\s]*)([0-9]+([kK])([—–\-~至])+[0-9]+([kK]))/;
  let salary = (text || '').match(regExp)?.[0] || '';
  if (!salary) {
    // 15k-25k
    regExp = /\b([0-9]+([kK])([—–\-~至])+[0-9]+([kK]))/;
    salary = (text || '').match(regExp)?.[0] || '';
  }
  return salary;
}

export function trackLinks(text: string) {
  const urls =
    (text || '').match(
      /(https?:\/\/)?([@a-z0-9.-]+)\.(com|cn|io|org|net|cc|info|biz|co|ai)\b\/?(?:\/[+~%/.\w-_]*)?\??(?:[-+=&;%@.\w_]*)*/g
    ) || [];
  return Array.from(
    new Set(urls.filter((url) => !url.includes('@')))
  ).map((url) => (url.endsWith('/') ? url.slice(0, url.length - 1) : url));
}

export function getScoreMap(configs: Config) {
  return configs
    .map((k) =>
      k.children.map((w) => {
        let alias: { name: string; score: number }[] = [];
        if (Array.isArray(w.alias)) {
          alias = w.alias.map((a) => ({ name: a, score: w.score }));
        } else if (typeof w.alias === 'string' && w.alias.length > 0) {
          alias = [{ name: w.alias, score: w.score }];
        }
        return [{ name: w.name, score: w.score }, ...alias];
      })
    )
    .flat(3)
    .reduce((result, it) => ({ ...result, [it.name]: it.score }), {});
}

const sentiment = new Sentiment();

export function calcSentiment(text: string, config: Config) {
  const textOk = (text || '').replace(/(\b[\w-]+\b)/g, ' $1 ');
  const result = sentiment.analyze(textOk, { extras: getScoreMap(config) });
  return result.comparative;
}

export function getBlogByLink(link: string) {
  // https://juejin.im/user/3403743728515246/
  // https://www.cnblogs.com/jiujuan/
  // https://blog.csdn.net/weixin_42134789
  // https://zhuanlan.zhihu.com/269840126
  // https://www.zhihu.com/people/gui-zi-29/posts
  // https://www.jianshu.com/u/c7757daadf27
  // https://segmentfault.com/u/jmxiao
  // https://my.oschina.net/u/267941
  // https://www.imooc.com/u/1215284
  // https://blog.51cto.com/12131824
  // https://cloud.tencent.com/developer/user/1307425
  // https://yq.aliyun.com/users/mgzlkjdaehkdc
  // https://jiayili.gitbooks.io/fe-study-easier/content/
  // https://github.com/guissy?tab=repositories
  // https://github.com/guissy/cc
  const csdn = /blog\.csdn\.net\/[A-z0-9_-]+/;
  const cnblogs = /www\.cnblogs\.com\/[A-z0-9_-]+/;
  const juejin = /juejin\.im\/user\/[A-z0-9_-]+/;
  const zhihuUser = /www\.zhihu\.com\/people\/[A-z0-9_-]+/;
  const zhihuZhuan = /zhuanlan\.zhihu\.com\/[A-z0-9_-]+/;
  const jianshu = /www\.jianshu\.com\/u\/[A-z0-9_-]+/;
  const segmentfault = /segmentfault\.com\/u\/[A-z0-9_-]+/;
  const oschina = /my\.oschina\.net\/u\/[A-z0-9_-]+/;
  const imooc = /www\.imooc\.com\/u\/[A-z0-9_-]+/;
  const cto = /blog\.51cto\.com\/[A-z0-9_-]+/;
  const tencent = /cloud\.tencent\.com\/developer\/user\/[A-z0-9_-]+/;
  const aliyun = /yq\.aliyun\.com\/users\/[A-z0-9_-]+/;
  const gitbooks = /jiayili\.gitbooks\.io\/[A-z0-9_-]+/;
  const github = /github\.com\/[A-z0-9_-]+/;
  const gitee = /gitee\.com\/[A-z0-9_-]+/;
  const n = [
    csdn,
    cnblogs,
    juejin,
    zhihuUser,
    zhihuZhuan,
    jianshu,
    segmentfault,
    oschina,
    imooc,
    cto,
    tencent,
    aliyun,
    gitbooks,
    github,
    gitee,
  ].findIndex((regexp) => regexp.test(link));
  const names = 'csdn,cnblogs,juejin,zhihu,zhihu,jianshu,segmentfault,oschina,imooc,51cto,tencent,aliyun,gitbooks,github,gitee'.split(
    ','
  );
  return n >= 0 ? names[n] : '';
}

export function getGithubByLink(links: string[]) {
  const github = /(github|giteeee)\.com\/[A-z0-9_-]+/g;
  return Array.from(
    new Set(
      links
        .map((link) => link.match(github)?.[0] || '')
        .filter(Boolean)
        .map((link) => `https://${link}`)
    )
  );
}

export function calcLevelSalaryRate(
  salary: string,
  levelValue: number
): number {
  const salaryRange = salary.match(/\d+/g) || [];
  const salaryMid =
    salaryRange.map((v) => parseInt(v, 10)).reduce((t, v) => t + v, 0) / 2;
  const rate = salaryMid > 0 && levelValue > 2 ? levelValue / salaryMid : 0;
  return rate / 0.29;
}

export function calcLevelSalaryStar(
  levelSalary: number,
  levelValue: number
): number {
  return (
    Math.round(Math.log2((levelSalary || 0.4) * 2.5) * levelValue * 0.62) / 2
  );
}
