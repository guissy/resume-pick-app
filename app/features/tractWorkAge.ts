export default function trackWorkAge(text: string) {
  // 工作年限：3年
  let years: string;
  const match1 = text.match(
    /(工作)?(年限|经验|开发)[^\d\n]*([\d一二三四五六七八九十]+)\s*年/
  );
  years = match1 ? String(match1.pop()) : '';
  if (!years) {
    const match2 = text.match(/([\d一二三四五六七八九十]+)\s*年(工作)?(经验)/);
    years = match2 ? String(match2[1]) : '';
  }
  return years ? `${years}年` : '-';
}

export function trackPhone(text: string) {
  return text.match(/1\d{10}/g)?.[0];
}
