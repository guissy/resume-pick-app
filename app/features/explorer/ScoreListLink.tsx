import React, { MouseEvent } from 'react';
import styles from './ScoreList.css';
import { getBlogByLink } from '../parser/tractWorkAge';
import Image from '../icon/image';

type Props = {
  link: string;
  onClick: (e: MouseEvent, link: string) => void;
};
const ScoreListLink: React.FC<Props> = ({ link, onClick }) => {
  return (
    <a
      href={link}
      className={styles.alink}
      style={{
        color: getBlogByLink(link) ? '#f1f1f1' : '#999999',
      }}
      onClick={(e) => onClick(e, link)}
    >
      <img
        className={styles.blogIcon}
        src={Image[`${getBlogByLink(link)}_png` as keyof typeof Image]}
        alt={getBlogByLink(link)}
      />
      {link.replace(/(https?:\/\/|www\.)/, '')}
    </a>
  );
};

export default ScoreListLink;
