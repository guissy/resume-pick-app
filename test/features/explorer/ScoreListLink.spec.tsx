import React from 'react';
import Enzyme, { mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import renderer from 'react-test-renderer';
import ScoreListLink from '../../../app/features/explorer/ScoreListLink';
import * as tractWorkAge from '../../../app/features/parser/tractWorkAge';

Enzyme.configure({ adapter: new Adapter() });

const defaultTestProps = { link: 'http://test.com', onClick: () => {} };

function setup(props = defaultTestProps) {
  // eslint-disable-next-line react/jsx-props-no-spreading
  const getWrapper = () => mount(<ScoreListLink {...props} />);
  const component = getWrapper();
  return {
    component,
  };
}
describe('ScoreListLink component', () => {
  it('should match exact snapshot', () => {
    const tree = renderer
      // eslint-disable-next-line react/jsx-props-no-spreading
      .create(<ScoreListLink {...defaultTestProps} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should call getBlogByLink', () => {
    const getBlogByLinkSpy = jest.spyOn(tractWorkAge, 'getBlogByLink');
    setup();
    expect(getBlogByLinkSpy).toBeCalledTimes(1);
    expect(getBlogByLinkSpy).toBeCalledWith(defaultTestProps.link);
  });

  it('should call onClick', () => {
    const onClick = jest.fn();
    const { component } = setup({ ...defaultTestProps, onClick });
    component.simulate('click');
    const mouseEvent = expect.objectContaining({ type: 'click' });
    expect(onClick).toBeCalledWith(mouseEvent, defaultTestProps.link);
  });
});
