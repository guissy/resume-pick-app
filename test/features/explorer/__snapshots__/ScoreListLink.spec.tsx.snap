// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScoreListLink component should match exact snapshot 1`] = `
<a
  className="alink"
  href="http://test.com"
  onClick={[Function]}
  style={
    Object {
      "color": "#999999",
    }
  }
>
  <img
    alt=""
    className="blogIcon"
  />
  test.com
</a>
`;
